### ISSUE TITLE
[JENKINS-35821] API to return per node averages, and ETA based on history of execution

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#2789]
<p><b>In scope</b></p>
<ul>
	<li>Data to support <a href="https://issues.jenkins.io/browse/JENKINS-40929" title="Developer can see durations when hovering on indicators" class="issue-link" data-issue-key="JENKINS-40929"><del>JENKINS-40929</del></a> but provide averages</li>
</ul>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay">jamesdumay</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821" target="_blank">API to return per node averages, and ETA based on history of execution</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Minor
<li><b>component(s)</b>: blueocean-plugin
<li><b>label(s)</b>: api, blueocean-imported
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 1
<li><b>watchers</b>: 3
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p><b>In scope</b></p>
<ul>
	<li>Data to support <a href="https://issues.jenkins.io/browse/JENKINS-40929" title="Developer can see durations when hovering on indicators" class="issue-link" data-issue-key="JENKINS-40929"><del>JENKINS-40929</del></a> but provide averages</li>
</ul>
</pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261325&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261325">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Waiting on the metrics from <a href="https://docs.google.com/document/d/1eOsw7xezQM6g9uh727flL6haPKUQqOBxYZiQvRjcV8Q/edit#" class="external-link" target="_blank" rel="nofollow noopener">https://docs.google.com/document/d/1eOsw7xezQM6g9uh727flL6haPKUQqOBxYZiQvRjcV8Q/edit#</a> </p>
</pre>
</details>

</ul>
</details>

<p>Waiting on the metrics from <a href="https://docs.google.com/document/d/1eOsw7xezQM6g9uh727flL6haPKUQqOBxYZiQvRjcV8Q/edit#" class="external-link" target="_blank" rel="nofollow noopener">https://docs.google.com/document/d/1eOsw7xezQM6g9uh727flL6haPKUQqOBxYZiQvRjcV8Q/edit#</a> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261372&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261372">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>FYI the stage view does this by calculating stage averages based on historical (we had a discussion elsewhere on that, but lost to the sands of time). </p>
</pre>
</details>

</ul>
</details>

<p>FYI the stage view does this by calculating stage averages based on historical (we had a discussion elsewhere on that, but lost to the sands of time). </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261508&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261508">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mmccaskill" class="user-hover" rel="mmccaskill">mmccaskill</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> has some funky new API coming.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mmccaskill" class="user-hover" rel="mmccaskill">mmccaskill</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> has some funky new API coming.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261298&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261298">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a>  This doesn't need the <b>full</b> metrics functionality, it just needs to ability to extract the timing data, which will precede a fully generic metrics API (since basic timing information is needed for several pieces). </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a>  This doesn't need the <b>full</b> metrics functionality, it just needs to ability to extract the timing data, which will precede a fully generic metrics API (since basic timing information is needed for several pieces). </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261222&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261222">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Specifically, this depends on: OSS-961 and OSS-962 and <b>maybe</b> OSS-963</p>
</pre>
</details>

</ul>
</details>

<p>Specifically, this depends on: OSS-961 and OSS-962 and <b>maybe</b> OSS-963</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261443&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261443">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vpandey" class="user-hover" rel="vpandey">vpandey</a> consider this comment from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> <a href="https://cloudbees.atlassian.net/browse/UX-393?focusedCommentId=37472&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-37472" class="external-link" target="_blank" rel="nofollow noopener">https://cloudbees.atlassian.net/browse/UX-393?focusedCommentId=37472&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-37472</a></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vpandey" class="user-hover" rel="vpandey">vpandey</a> consider this comment from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> <a href="https://cloudbees.atlassian.net/browse/UX-393?focusedCommentId=37472&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-37472" class="external-link" target="_blank" rel="nofollow noopener">https://cloudbees.atlassian.net/browse/UX-393?focusedCommentId=37472&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-37472</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vivek.png" title="vivek's avatar" /> vivek:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261522&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261522">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> Following up from your comment in other ticket:</p>

<blockquote>
<p>Comment on this, to consider: parallel blocks (and stages containing them) may have more than one state, I.E. one branch may fail where another passes. How does one represent mixed states?</p></blockquote>

<p>In our current implementation we have:</p>

<ul>
	<li>All parallel blocks completes and all are success, stage result is SUCCESS and state FINISHED</li>
	<li>All parallel blocks completes and at least one fail  stage result is FAILURE and state FINISHED</li>
	<li>If one or more parallel block is still running, stage result is UNKNOWN and state RUNNING</li>
</ul>


<blockquote>
<p>Other point: NotExecuted stages/nodes – this can happen when resuming from a checkpoint (the flownodes are created preceding the checkpoint but marked with NotExecutedNodeAction).</p></blockquote>

<p>I don't think we are handling these at the moment. So NotExecuted is something thats going to be executed eventually? Assuming thats what it is, I would expect  the behavior on our end to be state being QUEUED and result UNKNOWN. </p>

<blockquote>
<p>Implementation point: try/catch blocks can generate steps with nodes marked as failed where the overall FLOW still succeeds (error was caught). See <a href="https://issues.jenkins-ci.org/browse/JENKINS-34212" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-34212</a>
Hopefully by mentioning these issues now, they will avoid tripping you up later (as they have cropped up in stage view and had or are HAVING to be handled).</p></blockquote>

<p>This behavior sounds correct. Why are we marking that step as failure? IMO, user is explicitly declaring try/catch, that tells pipeline machinery should not flag it as ErrorAction. Or maybe you need to declare error codes to interpret it as, ignore from blueocean point of view and treat it as success? Not sure how you guys are planning to fix it.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> Following up from your comment in other ticket:</p>

<blockquote>
<p>Comment on this, to consider: parallel blocks (and stages containing them) may have more than one state, I.E. one branch may fail where another passes. How does one represent mixed states?</p></blockquote>

<p>In our current implementation we have:</p>

<ul>
	<li>All parallel blocks completes and all are success, stage result is SUCCESS and state FINISHED</li>
	<li>All parallel blocks completes and at least one fail  stage result is FAILURE and state FINISHED</li>
	<li>If one or more parallel block is still running, stage result is UNKNOWN and state RUNNING</li>
</ul>


<blockquote>
<p>Other point: NotExecuted stages/nodes – this can happen when resuming from a checkpoint (the flownodes are created preceding the checkpoint but marked with NotExecutedNodeAction).</p></blockquote>

<p>I don't think we are handling these at the moment. So NotExecuted is something thats going to be executed eventually? Assuming thats what it is, I would expect  the behavior on our end to be state being QUEUED and result UNKNOWN. </p>

<blockquote>
<p>Implementation point: try/catch blocks can generate steps with nodes marked as failed where the overall FLOW still succeeds (error was caught). See <a href="https://issues.jenkins-ci.org/browse/JENKINS-34212" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-34212</a><br/>
Hopefully by mentioning these issues now, they will avoid tripping you up later (as they have cropped up in stage view and had or are HAVING to be handled).</p></blockquote>

<p>This behavior sounds correct. Why are we marking that step as failure? IMO, user is explicitly declaring try/catch, that tells pipeline machinery should not flag it as ErrorAction. Or maybe you need to declare error codes to interpret it as, ignore from blueocean point of view and treat it as success? Not sure how you guys are planning to fix it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261268&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261268">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vpandey" class="user-hover" rel="vpandey">vpandey</a> The parallel blocks handling seems reasonable (though I think it needs to define handling for other combinations of states in a similar fashion). </p>

<p>&gt; I don't think we are handling these at the moment. So NotExecuted is something thats going to be executed eventually?</p>

<p>No, NotExecuted is set on all steps skipped when you resume from a checkpoint - those steps are <b>never</b> executed (since the checkpoint resume takes care of their results).   It's a separate state (stage view has handling for this), and it needs to be handled if we are going to support the CJP value-add in cloudbees workflow plugin.</p>

<p>&gt; This behavior sounds correct. Why are we marking that step as failure? IMO, user is explicitly declaring try/catch, that tells pipeline machinery should not flag it as ErrorAction. </p>

<p>That was my impression too - if the error is caught, the step should either not get an ErrorAction, or it should have some marker to indicate that it the error was caught.  Of course there is more than one school of thought on this matter unfortunately.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vpandey" class="user-hover" rel="vpandey">vpandey</a> The parallel blocks handling seems reasonable (though I think it needs to define handling for other combinations of states in a similar fashion). </p>

<p>> I don't think we are handling these at the moment. So NotExecuted is something thats going to be executed eventually?</p>

<p>No, NotExecuted is set on all steps skipped when you resume from a checkpoint - those steps are <b>never</b> executed (since the checkpoint resume takes care of their results).   It's a separate state (stage view has handling for this), and it needs to be handled if we are going to support the CJP value-add in cloudbees workflow plugin.</p>

<p>> This behavior sounds correct. Why are we marking that step as failure? IMO, user is explicitly declaring try/catch, that tells pipeline machinery should not flag it as ErrorAction. </p>

<p>That was my impression too - if the error is caught, the step should either not get an ErrorAction, or it should have some marker to indicate that it the error was caught.  Of course there is more than one school of thought on this matter unfortunately.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261344&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261344">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> hey mate whats the current status of the work on your end?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> hey mate whats the current status of the work on your end?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261428&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261428">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a> It's coming along, the core that I was finishing previously is done, and I'm got work well underway on the block scanning visitor.  Vivek and I will probably want to touch base next week about the API structure and consumption patterns.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a> It's coming along, the core that I was finishing previously is done, and I'm got work well underway on the block scanning visitor.  Vivek and I will probably want to touch base next week about the API structure and consumption patterns.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261452&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261452">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> fantastic. Could you setup a hangout with him next week?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> fantastic. Could you setup a hangout with him next week?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=261243&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261243">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a> Already done!</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a> Already done!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=271484&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-271484">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> is this really redundant now that the real work is to move to bismuth api? if so, feel free to close it. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> is this really redundant now that the real work is to move to bismuth api? if so, feel free to close it. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vivek.png" title="vivek's avatar" /> vivek:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=272206&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-272206">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> Bismuth API gives basic structure to parse executed flow nodes in SAX like parser, it also gives structure for timing info. Lets keep this ticket open as we need to implement these functionalities in blueocean api.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> Bismuth API gives basic structure to parse executed flow nodes in SAX like parser, it also gives structure for timing info. Lets keep this ticket open as we need to implement these functionalities in blueocean api.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=274697&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-274697">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> is there an API todo this yet?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> is there an API todo this yet?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=274827&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-274827">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>As discussed:</p>
<ul>
	<li>There is a low level API in bismuth to retrieve this data.</li>
	<li>We will use the low level API in Blue Ocean to drive out the requirements of a future higher level API</li>
	<li>Think of it as a PoC that lives within Blue Ocean that we could either throw away or reuse.</li>
	<li><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> to meet with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> to co-ordinate how it should be done.</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>As discussed:</p>
<ul>
	<li>There is a low level API in bismuth to retrieve this data.</li>
	<li>We will use the low level API in Blue Ocean to drive out the requirements of a future higher level API</li>
	<li>Think of it as a PoC that lives within Blue Ocean that we could either throw away or reuse.</li>
	<li><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> to meet with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a> to co-ordinate how it should be done.</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35821?focusedId=276245&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-276245">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Per meeting with @vivek on Friday, there are a couple parts to this, and by breaking it up into smaller pieces we can make it easier.  We have a followup meeting planned in a couple weeks to synch again and hammer out more of the fine details. </p>

<p>Pieces:</p>

<ol>
	<li>Collect a bit more structural information in BO during flow analysis (an additional Map or two).  This lets us overlay a treelike, or DOM-like overlay of flow node data (depth-limited for BO though, so simplified).</li>
	<li>Create a generic API accepting DOM-like interface, which generates mappings of similar ("homologous") parts of the flow execution across WorkflowRuns (i.e. stage 'bob' in Run #1 maps to stage 'bob' in Run #2).</li>
</ol>
<ul>
	<li>This combines a mapping component (first stage with same name matches) with a filtering component (ex: if stage didn't complete with SUCCESSFUL, it can't be used for prediction).</li>
	<li>This will support pluggable strategies to accomplish this (heuristics).  This lets us do the simplest version then rip it out entirely if needed (and give Blue Ocean a basic form without much nesting, but analytics can do a fancier one if desired).  Also important because mappings are very fiddly (see: tons of Stage View Bugs logged for this aspect).</li>
	<li>Mappings will be fuzzy/best-case, and can fail entirely if two runs are too different.</li>
	<li>For more complex mappings we'll do it recursively to simplify this &#8211; map stages against each other, then map steps within each stage.</li>
</ul>
<ol>
	<li>Take the homologous (similar) pieces of flows and generate predictions for run time by aggregating them.  We can use the status/timing APIs.   This will combine run time, pause time, and maybe status.  It should be VERY simple by design.  Probably just do an average or median of times (subtracting pause), maybe report an error bar if enough of them.</li>
	<li>Optional: a basic API where you can request run analysis by giving a WorkflowRun as input &#8211; this can internally use caching or an analysis thread pool (to avoid overloading the system).</li>
</ol>


<p>Basically what we'd do is digest a flow, then digest previous runs, try to map similar parts, then see if we have enough data for predictions, find the similar bits, then do estimates based on those.   We're aiming for the simplest workable solution initially. </p>

<p>Note: to map flow chunks, <b>each flowchunk must be identified by</b>: </p>

<ol>
	<li>Label if present (for stages, parallel branches).</li>
	<li>Index within enclosing container (example, 1st FlowNode in the stage, 1st FlowNode in a parallel, 1st stage in a run).</li>
	<li>EITHER containers have a list of contained chunks (i.e. for BO a WorkflowRun will have a list of stages and parallel blocks), OR each chunk lists its parent container.  This is to let us (for example) see if we have 2 parallel blocks with the same branch names vs. one parallel block.</li>
</ol>


<p>*<b>Why all this?</b>*</p>

<ol>
	<li>We're seeing increasingly dynamic pipeline structures &#8211; conditional execution of parallel branches, whole stages, or more.</li>
	<li>We can't directly line up nodes by ID, because some stages may have different numbers of steps (example: retry blocks).</li>
	<li>Generating mappings has all the hard bits, so we isolate it for testability and simplicity.</li>
	<li>We can delegate to the pipeline graph analysis StatusAndTiming APIs for all the timing info</li>
</ol>

</pre>
</details>

</ul>
</details>

<p>Per meeting with @​vivek on Friday, there are a couple parts to this, and by breaking it up into smaller pieces we can make it easier.  We have a followup meeting planned in a couple weeks to synch again and hammer out more of the fine details. </p>

<p>Pieces:</p>

<ol>
	<li>Collect a bit more structural information in BO during flow analysis (an additional Map or two).  This lets us overlay a treelike, or DOM-like overlay of flow node data (depth-limited for BO though, so simplified).</li>
	<li>Create a generic API accepting DOM-like interface, which generates mappings of similar ("homologous") parts of the flow execution across WorkflowRuns (i.e. stage 'bob' in Run #1 maps to stage 'bob' in Run #2).</li>
</ol>
<ul>
	<li>This combines a mapping component (first stage with same name matches) with a filtering component (ex: if stage didn't complete with SUCCESSFUL, it can't be used for prediction).</li>
	<li>This will support pluggable strategies to accomplish this (heuristics).  This lets us do the simplest version then rip it out entirely if needed (and give Blue Ocean a basic form without much nesting, but analytics can do a fancier one if desired).  Also important because mappings are very fiddly (see: tons of Stage View Bugs logged for this aspect).</li>
	<li>Mappings will be fuzzy/best-case, and can fail entirely if two runs are too different.</li>
	<li>For more complex mappings we'll do it recursively to simplify this &#8211; map stages against each other, then map steps within each stage.</li>
</ul>
<ol>
	<li>Take the homologous (similar) pieces of flows and generate predictions for run time by aggregating them.  We can use the status/timing APIs.   This will combine run time, pause time, and maybe status.  It should be VERY simple by design.  Probably just do an average or median of times (subtracting pause), maybe report an error bar if enough of them.</li>
	<li>Optional: a basic API where you can request run analysis by giving a WorkflowRun as input &#8211; this can internally use caching or an analysis thread pool (to avoid overloading the system).</li>
</ol>


<p>Basically what we'd do is digest a flow, then digest previous runs, try to map similar parts, then see if we have enough data for predictions, find the similar bits, then do estimates based on those.   We're aiming for the simplest workable solution initially. </p>

<p>Note: to map flow chunks, <b>each flowchunk must be identified by</b>: </p>

<ol>
	<li>Label if present (for stages, parallel branches).</li>
	<li>Index within enclosing container (example, 1st FlowNode in the stage, 1st FlowNode in a parallel, 1st stage in a run).</li>
	<li>EITHER containers have a list of contained chunks (i.e. for BO a WorkflowRun will have a list of stages and parallel blocks), OR each chunk lists its parent container.  This is to let us (for example) see if we have 2 parallel blocks with the same branch names vs. one parallel block.</li>
</ol>


<p>*<b>Why all this?</b>*</p>

<ol>
	<li>We're seeing increasingly dynamic pipeline structures &#8211; conditional execution of parallel branches, whole stages, or more.</li>
	<li>We can't directly line up nodes by ID, because some stages may have different numbers of steps (example: retry blocks).</li>
	<li>Generating mappings has all the hard bits, so we isolate it for testability and simplicity.</li>
	<li>We can delegate to the pipeline graph analysis StatusAndTiming APIs for all the timing info</li>
</ol>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-35849%22">JENKINS-35849</a>]</i>
